### PR TITLE
Feature/implement machine health checker

### PR DESCRIPTION
**Pull request by wongni**
*Issue #, if available:*

*Description of changes:*

*Testing performed:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: api/v1beta1/cloudstackmachinestatechecker_types.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// CloudStackMachineStateCheckerSpec defines the desired state of CloudStackMachineStateChecker
+type CloudStackMachineStateCheckerSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
```
Comment (rejoshed) [inline api/v1beta1/cloudstackmachinestatechecker_types.go:28]: Like these. : )

File: controllers/cloudstackmachinestatechecker_controller.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	csCtrlrUtils "github.com/aws/cluster-api-provider-cloudstack/controllers/utils"
+	capiv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackmachinestatecheckers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackmachinestatecheckers/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackmachinestatecheckers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines,verbs=get;list;watch;delete
+
+// CloudStackMachineStateCheckerReconciliationRunner is a ReconciliationRunner with extensions specific to CloudStackCluster reconciliation.
```
Comment (rejoshed) [inline controllers/cloudstackmachinestatechecker_controller.go:35]: `specific to CloudStackCluster`

Please revise to reflect CRD.

File: controllers/cloudstackmachinestatechecker_controller.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	}
+	if r.CSMachine.Status.InstanceState == "Running" {
+		r.ReconciliationSubject.Status.Ready = true
+	} else {
+		if err := r.Client.Delete(r.RequestCtx, r.CAPIMachine); err != nil {
+			return r.ReturnWrappedError(err, "failed to delete CAPI machine")
+		}
+	}
+
+	return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
```
Comment (rejoshed) [inline controllers/cloudstackmachinestatechecker_controller.go:87]: Should we make the requeue period customizable?

File: controllers/cloudstackmachinestatechecker_controller.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		return res, err
+	}
+	if res, err := r.GetParent(r.CSMachine, r.CAPIMachine)(); r.ShouldReturn(res, err) {
+		return res, err
+	}
+
+	if err := r.CS.ResolveVMInstanceDetails(r.CSMachine); err != nil {
+		return r.ReturnWrappedError(err, "failed to resolve VM instance details")
+	}
+	if r.CSMachine.Status.InstanceState == "Running" {
```
Comment (rejoshed) [inline controllers/cloudstackmachinestatechecker_controller.go:79]: Easy peasy! Love it!

File: controllers/cloudstackzone_controller.go [hunk 5]
```diff
@@ -37,7 +36,6 @@ type CloudStackZoneReconciliationRunner struct {
 	csCtrlrUtils.ReconciliationRunner
 	Zones                 *infrav1.CloudStackZoneList
 	ReconciliationSubject *infrav1.CloudStackZone
-	CSUser                cloud.Client
```
Comment (rejoshed) [inline controllers/cloudstackzone_controller.go:40]: This will be used for account and domain switching. No need to remove it.

**Review by rejoshed — COMMENTED** [review]
Overall, looks pretty good. Guess we need to cascade this merge post my CRD PR.

Also, we had discussed adding E2E tests for new functionality. This probably fits the bill. Also I grok the controller test suite a lot better now than I used to, and I'd be happy to help you add a controller test to this too.

I wouldn't say these are blockers though.

**Comment by jweite-amazon** [comment]
Approved, pending mergability and tests passing.

File: api/v1beta1/cloudstackmachinestatechecker_types.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
[the remark this answers is not in the document]
Comment (wongni) [reply api/v1beta1/cloudstackmachinestatechecker_types.go:23]: Done

File: api/v1beta1/cloudstackmachinestatechecker_types.go (hunk 1 above)
Comment (wongni) [reply api/v1beta1/cloudstackmachinestatechecker_types.go:28]: Done

File: controllers/cloudstackmachinestatechecker_controller.go (hunk 2 above)
Comment (wongni) [reply controllers/cloudstackmachinestatechecker_controller.go:35]: Done

File: controllers/cloudstackmachinestatechecker_controller.go (hunk 3 above)
Comment (wongni) [reply controllers/cloudstackmachinestatechecker_controller.go:87]: Let's leave it as is. When we have a request to make the period customizable let's revisit this.

File: api/v1beta1/cloudstackmachinestatechecker_types.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CloudStackMachineStateCheckerSpec defines the desired state of CloudStackMachineStateChecker
+type CloudStackMachineStateCheckerSpec struct {
+	// Give better description.
```
Comment (rejoshed) [inline api/v1beta1/cloudstackmachinestatechecker_types.go:25]: @Wonkun! Haha

Comment (rejoshed) [reply api/v1beta1/cloudstackmachinestatechecker_types.go:25]: Also the comment above it no longer makes sense really. Please update that as well.

File: controllers/cloudstackaffinitygroup_controller.go [hunk 8]
```diff
@@ -33,7 +33,7 @@ import (
 // Need to watch machine templates for creation of an affinity group.
 //+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackmachinetemplate,verbs=get;list;watch;
 
-// CloudStackAGReconciliationRunner is a ReconciliationRunner with extensions specific to CloudStackCluster reconciliation.
+// CloudStackAGReconciliationRunner is a ReconciliationRunner with extensions specific to CloudStack affinity group reconciliation.
```
Comment (rejoshed) [inline controllers/cloudstackaffinitygroup_controller.go:36]: Nice catch.

**Review by rejoshed — APPROVED** [review]

**Pull request merged**